zenith_utils/src/sync.rs

```diff
@@ -0,0 +1,133 @@
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::{io, task};
+
+pin_project! {
+    /// We use this future to mark certain methods
+    /// as callable in both sync and async modes
+    #[repr(transparent)]
+    pub struct SyncFuture<S, T: Future> {
+        #[pin]
+        nested: T,
+        _marker: std::marker::PhantomData<S>,
+    }
+}
+
+impl<S, T: Future> SyncFuture<S, T> {
+    pub(crate) fn new(nested: T) -> Self {
+        Self {
+            nested,
+            _marker: Default::default(),
+        }
+    }
+}
+
+impl<S, T: Future> Future for SyncFuture<S, T> {
+    type Output = T::Output;
+
+    #[inline(always)]
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> task::Poll<Self::Output> {
+        self.project().nested.poll(cx)
+    }
+}
+
+/// Proves that we can call [SyncFuture::wait]
+pub trait SyncProof {}
+
+impl<T: SyncProof> SyncProof for &T {}
+impl<T: SyncProof> SyncProof for &mut T {}
+
+impl<P: SyncProof, T: Future> SyncFuture<P, T> {
+    /// Synchronously wait for future completion
+    pub fn wait(mut self) -> T::Output {
+        const RAW_WAKER: task::RawWaker = task::RawWaker::new(
+            std::ptr::null(),
+            &task::RawWakerVTable::new(
+                |_| RAW_WAKER,
+                |_| panic!("SyncFuture: failed to wake"),
+                |_| panic!("SyncFuture: failed to wake by ref"),
+                |_| { /* drop is no-op */ },
+            ),
+        );
+
+        // SAFETY: We never move `self` during this call;
+        // furthermore, it will be dropped in the end regardless of panics
+        let this = unsafe { Pin::new_unchecked(&mut self) };
+
+        // SAFETY: This waker doesn't do anything apart from panicking
+        let waker = unsafe { task::Waker::from_raw(RAW_WAKER) };
+        let context = &mut task::Context::from_waker(&waker);
+
+        match this.poll(context) {
+            task::Poll::Ready(res) => res,
+            _ => panic!("SyncFuture: unexpected pending!"),
+        }
+    }
+}
+
+/// With this, we can turn any `Read` into a blocking `AsyncRead`
+#[repr(transparent)]
+pub struct AsyncishRead<T: io::Read + Unpin>(pub T);
+
+impl<T: io::Read + Unpin> SyncProof for AsyncishRead<T> {}
+
+impl<T: io::Read + Unpin> tokio::io::AsyncRead for AsyncishRead<T> {
+    #[inline(always)]
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        _cx: &mut task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> task::Poll<io::Result<()>> {
+        task::Poll::Ready(
+            // This will block, meaning we don't need a real event loop!
+            self.0
```

Comment:
Could you elaborate? Why this would block? Will it block when called from async context?

Reply:
It will block regardless of the context because we call `Read::read`, which is inherently synchronous.

Reply:
Then it seems that there is no benefit in async execution, right?

Reply:
> Then it seems that there is no benefit in async execution, right?

Not quite. You can still use `read_fut`, which is fully asynchronous, as it doesn't use `AsyncishRead`.

Reply:
To elaborate: `SyncFuture` is a future which we may synchronously wait for **under certain circumstances**. AsyncishRead serves as a proof that we may do that. In async context, we may always just `.await` it.